lib/quarantine.rb

```diff
@@ -45,84 +44,61 @@ def database
       end
   end
 
-  # Scans the quarantine_list from the database and store their IDs in quarantined_ids
-  def fetch_quarantine_list
+  # Scans the test_statuses from the database and store their IDs in quarantined_ids
+  def fetch_test_statuses
     begin
-      quarantine_list = database.scan(options[:list_table])
+      test_statuses = database.scan(options[:test_statuses_table_name])
     rescue Quarantine::DatabaseError => e
-      add_to_summary(:database_failures, "#{e&.cause&.class}: #{e&.cause&.message}")
+      @database_failures << "#{e&.cause&.class}: #{e&.cause&.message}"
       raise Quarantine::DatabaseError.new(
         <<~ERROR_MSG
-          Failed to pull the quarantine list from #{options[:list_table]}
+          Failed to pull the quarantine list from #{options[:test_statuses_table_name]}
           because of #{e&.cause&.class}: #{e&.cause&.message}
         ERROR_MSG
       )
     end
 
-    @quarantined_ids = quarantine_list.map { |q| q['id'] }
+    @quarantined_ids = test_statuses
+                       .group_by { |t| t['id'] }
+                       .map { |_id, tests| tests.max_by { |t| t['created_at'] } }
+                       .filter { |t| t['last_status'] == 'quarantined' }
+                       .map { |t| t['id'] }
   end
 
-  # Based off the type, upload a list of tests to a particular database table
-  def upload_tests(type)
-    case type
-    when :failed
-      tests = failed_tests
-      table_name = options[:failed_tests_table]
-    when :flaky
-      tests = flaky_tests
-      table_name = options[:list_table]
-    else
-      raise Quarantine::UnknownUploadError.new(
-        "Quarantine gem did not know how to handle #{type} upload of tests to dynamodb"
-      )
-    end
-
-    return unless tests.length < 10 && tests.length > 0
+  def upload_tests
+    return if tests.empty? || tests.values.count { |test| test.status == :quarantined } >= 10
```

Comment:
Several months ago, Flexport introduced a bug into our CI setup that caused _every_ test to become flaky, so this gem dutifully quarantined every test. While this made CI  run significantly faster, we decided to prevent it from happening again by adding a failsafe: since 10 tests suddenly becoming flaky in a single run is highly unlikely, it indicates something has gone wrong, so this logic bails out.

They picked (and harcoded) the number 10 just to get the fix out the door quickly, I think. I make it configurable in [the next PR](https://github.com/flexport/quarantine/pull/25).

Reply:
👍